src/backend/utils/resgroup/resgroup.c

```diff
@@ -4766,6 +4828,81 @@ moveQueryCheck(int sessionId, Oid groupId)
 	cdbdisp_clearCdbPgResults(&cdb_pgresults);
 }
 
+/*
+ * Try to give away all slot control to target process.
+ */
+static void
+resGroupGiveSlotAway(int sessionId, ResGroupSlotData **slot, Oid groupId)
+{
+	long timeout;
+	int64 curTime;
+	int64 waitStart;
+	int latchRes;
+	bool clean = false;
+	bool res = false;
+	
+	SIMPLE_FAULT_INJECTOR("resource_group_give_away1");
+
+	if (!ResGroupMoveSignalTarget(sessionId, *slot, groupId, false))
+		elog(ERROR, "can't send signal to process");
+	
+	waitStart = GetCurrentIntegerTimestamp();
+
+	for (;;)
+	{
+		CHECK_FOR_INTERRUPTS();
+
+		curTime = GetCurrentIntegerTimestamp();
+		timeout = gp_resource_group_move_timeout - (curTime - waitStart) / 1000;
+		if (timeout > 0)
+		{
+			PG_TRY();
+			{		
+				SIMPLE_FAULT_INJECTOR("resource_group_give_away2-1");
+				SIMPLE_FAULT_INJECTOR("resource_group_give_away2-2");
+				
+				latchRes = WaitLatch(&MyProc->procLatch,
+					WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH, timeout);
+
+				if (latchRes & WL_POSTMASTER_DEATH)
+					elog(ERROR,
+						"got WL_POSTMASTER_DEATH waiting on latch; exiting...");
+			}
+			PG_CATCH();
+			{
+				clean = true;
+				ResGroupMoveCheckTargetReady(sessionId, &clean, &res);
+				if (res)
+				{
+					/*
+					 * clean slot variable, because we don't need to touch it
+					 * in current process as control is on the target side
+					 */
+					*slot = NULL;
+					elog(WARNING,
+						"got exception, but slot control is on the target process side; not moving on QEs...");
+				}
+				PG_RE_THROW();
+			}
+			PG_END_TRY();
+		}
+		else
+			latchRes = WL_TIMEOUT;
+
+		SIMPLE_FAULT_INJECTOR("resource_group_give_away3");
+
+		clean = (latchRes & WL_TIMEOUT);
+		ResGroupMoveCheckTargetReady(sessionId, &clean, &res);
+		if (clean)
+			break;
+
+		ResetLatch(&MyProc->procLatch);
+	}
+
+	if (!res)
+		elog(ERROR, "target process can't move itself to a new group");
```

Comment:
I don't see any big difference, but it's shorter at least. OK.

Reply:
I should explain: because of "can't" I thought you mean it is an  _invalid_, not failed operation, as sentence is quite similar to https://github.com/arenadata/gpdb/blob/665710722804bda1ad7248a8214c70f6ee29dd06/gpcontrib/gp_internal_tools/gp_resource_group.c#L87-L90

~dushnila~

Reply:
And I guess it is better to use `ereport` in the new code.

Reply:
Still can't understand the play around "can't".  Quote: "Can’t is a contraction of cannot...", so they are the same.
There is a thing which may be interesting and makes sense: "...and as such it’s sometimes unsuitable for formal writing.", but it seems that's not what you meant.

Reply:
Moved from "can't" to "cannot" to be more formal.
`elog` replaced by "ereport" where it was possible.